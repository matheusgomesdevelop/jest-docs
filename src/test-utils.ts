import axios from "axios";

export const sum = (num1: number, num2: number) => num1 + num2;

// Errors
export const errorMessage = "Deu um erro de exceção";

export const dispatchError = () => {
  throw new Error(errorMessage);
};

// Promises
export const fetchData = () => {
  return new Promise((resolve, reject) => {
    const data = "peanut butter";
    resolve(data);
  });
};

// Initialize city databases
export const initializeCityDatabase = () => {
  const CITY_DATABASE = ["São vicente", "Praia Grande", "Santos", "Guarujá"];

  return CITY_DATABASE;
};

export const isCity = (cityList: Array<string>, name: string) => {
  return cityList.includes(name);
};

// Usando uma função de simulação

export const util_callback = (x: number) => 42 + x;
export function forEach(items: Array<number>, callback: (x: number) => void) {
  for (let index = 0; index < items.length; index++) {
    callback(items[index]);
  }
}

// Users
export const users = async () => {
    const users = await axios.get("http://localhost:3000/users.json");

    return users.data;
};
