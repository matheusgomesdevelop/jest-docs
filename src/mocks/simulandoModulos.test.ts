const axios = require("axios");

import { users } from "../../src/test-utils";

import defaultExport, { bar, foo } from "../foo-bar-baz";

jest.mock("axios");
jest.mock("../foo");
jest.mock("../foo-bar-baz", () => {
  const originalModule = jest.requireActual("../foo-bar-baz");

  // Mock da exportação default e renomeando export 'foo'
  return {
    __esModule: true,
    ...originalModule,
    default: jest.fn(() => "mocked baz"),
    foo: "mocked foo",
  };
});

describe("Simulando modulos", () => {
  // ==============================================
  it("Deve mockar a resposta do axios", async () => {
    const users_data = [{ name: "Bob" }];
    const resp = { data: users_data };

    /*
     * Retorna os dados que queremos que nosso teste afirme contra.
     * Assim dizemos para o axios.get retornar uma resposta falsa.
     */

    axios.get.mockResolvedValue(resp);

    return users().then((data) => expect(data).toEqual(users_data));
  });

  // ==============================================

  it("Deve sobrescrever uma parte do modulo e manter o resto como original", () => {
    const defaultExportResult = defaultExport();

    expect(defaultExportResult).toBe("mocked baz");
    expect(defaultExport).toHaveBeenCalled();

    expect(foo).toBe("mocked foo");
    expect(bar()).toBe("bar");
  });

  // ==============================================

  it("Deve mudar o comportamento da implementação da função com mock", () => {
    const fooMock = require("../foo");

    // Muda o comportamento da função para retornar de uma nova maneira como você preferir
    fooMock.mockImplementation(() => 42);
    expect(fooMock()).toBe(42);

    // Para mudar o comportamento de funcoes mais complexas
    fooMock.mockImplementation(() => 42).mockImplementationOnce(() => 43);
    expect(fooMock()).toBe(43);
  });

  // ==============================================

  describe("Deve criar nomes personalizados para os mocks", () => {
    // ==============================================

    it("Mock names", () => {
      const myMockFn = jest
        .fn()
        .mockReturnValue("default")
        .mockImplementation((scalar) => 42 + scalar)
        .mockName("add42");
    });

    // ==============================================

    describe("Matches personalizados", () => {
      // ==============================================

      const myMockFn = jest
        .fn()
        .mockReturnValue("default")
        .mockImplementation((scalar) => 42 + scalar)
        .mockName("add42");

      // ==============================================
      myMockFn(7);
      myMockFn(3);

      it("A função foi chamada pelo menos uma vez.", () => {
        expect(myMockFn).toHaveBeenCalled();
      });

      it("A função foi chamada com o argumento: 7", () => {
        expect(myMockFn).toHaveBeenCalledWith(3);
      });

      it("A ultima chamada da função foi executada com o parameto: 3", () => {
        expect(myMockFn).toHaveBeenLastCalledWith(3);
      });

      it("A chamada e o nome foi escrita de acordo com o teste de snapshot", () => {
        expect(myMockFn).toMatchSnapshot();
      });

      it("O numero de chamadas do mock foi maior que o parametro passado", () => {
        expect(myMockFn.mock.calls.length).toBeGreaterThan(0);
      });

      it("A função mock foi chamada pelo menos uma vez com o parametro especifico", () => {
        expect(myMockFn.mock.calls).toContainEqual([3]);
      });

      it("A ultima chamada da funcao recebeu os parametros especificos", () => {
        expect(myMockFn.mock.calls[myMockFn.mock.calls.length - 1]).toEqual([3]);
      });

      it("Deve identificar o primeiro argumento da ultima chamada.", () => {
        expect(myMockFn.mock.calls[myMockFn.mock.calls.length - 1][0]).toBe(3);
      });

      it("Deve verificar o nome da funcao mock executada", () => {
        expect(myMockFn.getMockName()).toBe("add42");

        myMockFn.mockName("novoNome");

        expect(myMockFn.getMockName()).toBe("novoNome");
      });
    });
  });
});
