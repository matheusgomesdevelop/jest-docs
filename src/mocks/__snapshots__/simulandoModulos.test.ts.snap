// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Simulando modulos Deve criar nomes personalizados para os mocks Matches personalizados A chamada e o nome foi escrita de acordo com o teste de snapshot 1`] = `
[MockFunction add42] {
  "calls": [
    [
      7,
    ],
    [
      3,
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": 49,
    },
    {
      "type": "return",
      "value": 45,
    },
  ],
}
`;
